File: Level 4/python/bankingsystem.py
# banking_system.py

from account import Account
from transactions import DepositTransaction, TransferTransaction
import heapq

class BankingSystem:
    def __init__(self):
        self.accounts = {}  # account_id -> Account
        self.outgoing_totals = {}  # account_id -> int
        self.payment_counter = 0
        self.scheduled_cashbacks = []  # min-heap of (timestamp, account_id, amount, payment_id)
        self.payment_status = {}  # account_id -> {payment_id: status}
        self.account_history = {}  # account_id -> list of (timestamp, balance)
        self.merge_map = {}  # merged_id -> active_id

    def _resolve_account(self, account_id):
        while account_id in self.merge_map:
            account_id = self.merge_map[account_id]
        return account_id

    def create_account(self, timestamp: int, account_id: str) -> bool:
        if account_id in self.accounts:
            return False
        self.accounts[account_id] = Account(account_id)
        self.outgoing_totals[account_id] = 0
        self.payment_status[account_id] = {}
        self.account_history[account_id] = [(timestamp, 0)]
        return True

    def _update_balance_history(self, account_id, timestamp):
        resolved = self._resolve_account(account_id)
        balance = self.accounts[resolved].get_balance()
        self.account_history[resolved].append((timestamp, balance))

    def deposit(self, timestamp: int, account_id: str, amount: int) -> int | None:
        self._process_cashbacks(timestamp)
        resolved = self._resolve_account(account_id)
        if resolved not in self.accounts:
            return None
        self.accounts[resolved].deposit(amount)
        self._update_balance_history(resolved, timestamp)
        return self.accounts[resolved].get_balance()

    def transfer(self, timestamp: int, source_id: str, target_id: str, amount: int) -> int | None:
        self._process_cashbacks(timestamp)
        s_id, t_id = self._resolve_account(source_id), self._resolve_account(target_id)
        if s_id == t_id or s_id not in self.accounts or t_id not in self.accounts:
            return None
        if not self.accounts[s_id].can_withdraw(amount):
            return None
        self.accounts[s_id].withdraw(amount)
        self.accounts[t_id].deposit(amount)
        self.outgoing_totals[s_id] += amount
        self._update_balance_history(s_id, timestamp)
        self._update_balance_history(t_id, timestamp)
        return self.accounts[s_id].get_balance()

    def top_spenders(self, timestamp: int, n: int) -> list[str]:
        self._process_cashbacks(timestamp)
        merged_totals = {}
        for acc, total in self.outgoing_totals.items():
            active = self._resolve_account(acc)
            merged_totals[active] = merged_totals.get(active, 0) + total
        sorted_accounts = sorted(merged_totals.items(), key=lambda x: (-x[1], x[0]))
        return [f"{k}({v})" for k, v in sorted_accounts[:n]]

    def pay(self, timestamp: int, account_id: str, amount: int) -> str | None:
        self._process_cashbacks(timestamp)
        resolved = self._resolve_account(account_id)
        if resolved not in self.accounts:
            return None
        if not self.accounts[resolved].can_withdraw(amount):
            return None
        self.accounts[resolved].withdraw(amount)
        self.outgoing_totals[resolved] += amount
        self.payment_counter += 1
        payment_id = f"payment{self.payment_counter}"
        cashback_time = timestamp + 86400000
        cashback = amount * 2 // 100
        heapq.heappush(self.scheduled_cashbacks, (cashback_time, resolved, cashback, payment_id))
        self.payment_status[resolved][payment_id] = "IN_PROGRESS"
        self._update_balance_history(resolved, timestamp)
        return payment_id

    def get_payment_status(self, timestamp: int, account_id: str, payment_id: str) -> str | None:
        self._process_cashbacks(timestamp)
        resolved = self._resolve_account(account_id)
        if resolved not in self.accounts:
            return None
        return self.payment_status.get(resolved, {}).get(payment_id)

    def get_balance(self, timestamp: int, account_id: str, time_at: int) -> int | None:
        resolved = self._resolve_account(account_id)
        if resolved not in self.account_history:
            return None
        history = self.account_history[resolved]
        result = None
        for t, bal in history:
            if t <= time_at:
                result = bal
            else:
                break
        return result

    def merge_accounts(self, timestamp: int, acc1: str, acc2: str) -> bool:
        root1 = self._resolve_account(acc1)
        root2 = self._resolve_account(acc2)
        if root1 == root2 or root1 not in self.accounts or root2 not in self.accounts:
            return False
        self.accounts[root1].deposit(self.accounts[root2].get_balance())
        self.outgoing_totals[root1] += self.outgoing_totals.get(root2, 0)
        for pid, status in self.payment_status[root2].items():
            self.payment_status[root1][pid] = status
        for t, bal in self.account_history[root2]:
            self.account_history[root1].append((timestamp, self.accounts[root1].get_balance()))
        self.merge_map[root2] = root1
        del self.accounts[root2]
        del self.outgoing_totals[root2]
        del self.payment_status[root2]
        del self.account_history[root2]
        return True

    def _process_cashbacks(self, timestamp: int):
        while self.scheduled_cashbacks and self.scheduled_cashbacks[0][0] <= timestamp:
            _, acc_id, cashback, pid = heapq.heappop(self.scheduled_cashbacks)
            resolved = self._resolve_account(acc_id)
            if resolved in self.accounts:
                self.accounts[resolved].deposit(cashback)
                self.payment_status[resolved][pid] = "CASHBACK_RECEIVED"
                self._update_balance_history(resolved, _)

File: Level 4/C#/bankingSystem.cs
// Level 4 - Banking System with Account Merging and Historical Balance Tracking in C#
using System;
using System.Collections.Generic;
using System.Linq;

public class BankingSystem
{
    private readonly Dictionary<string, Account> _accounts = new();
    private readonly Dictionary<string, int> _outgoingTotals = new();
    private readonly Dictionary<string, Dictionary<string, string>> _paymentStatus = new();
    private readonly PriorityQueue<(long timestamp, string accountId, int cashback, string paymentId), long> _cashbackQueue = new();
    private readonly Dictionary<string, List<(long timestamp, int balance)>> _accountHistory = new();
    private readonly Dictionary<string, string> _mergeMap = new();
    private long _paymentCounter = 0;
    private const long MILLISECONDS_IN_A_DAY = 86400000;

    private string ResolveAccount(string accountId)
    {
        while (_mergeMap.ContainsKey(accountId))
        {
            accountId = _mergeMap[accountId];
        }
        return accountId;
    }

    public bool CreateAccount(int timestamp, string accountId)
    {
        if (_accounts.ContainsKey(accountId)) return false;

        _accounts[accountId] = new Account(accountId);
        _outgoingTotals[accountId] = 0;
        _paymentStatus[accountId] = new Dictionary<string, string>();
        _accountHistory[accountId] = new List<(long, int)> { (timestamp, 0) };
        return true;
    }

    private void UpdateBalanceHistory(string accountId, long timestamp)
    {
        string resolved = ResolveAccount(accountId);
        int balance = _accounts[resolved].GetBalance();
        _accountHistory[resolved].Add((timestamp, balance));
    }

    public int? Deposit(long timestamp, string accountId, int amount)
    {
        ProcessCashbacks(timestamp);
        string resolved = ResolveAccount(accountId);
        if (!_accounts.TryGetValue(resolved, out var account)) return null;

        account.Deposit(amount);
        UpdateBalanceHistory(resolved, timestamp);
        return account.GetBalance();
    }

    public int? Transfer(long timestamp, string sourceId, string targetId, int amount)
    {
        ProcessCashbacks(timestamp);
        string src = ResolveAccount(sourceId);
        string tgt = ResolveAccount(targetId);

        if (src == tgt || !_accounts.ContainsKey(src) || !_accounts.ContainsKey(tgt)) return null;

        var source = _accounts[src];
        var target = _accounts[tgt];

        if (!source.Withdraw(amount)) return null;

        target.Deposit(amount);
        _outgoingTotals[src] += amount;
        UpdateBalanceHistory(src, timestamp);
        UpdateBalanceHistory(tgt, timestamp);
        return source.GetBalance();
    }

    public string? Pay(long timestamp, string accountId, int amount)
    {
        ProcessCashbacks(timestamp);
        string resolved = ResolveAccount(accountId);
        if (!_accounts.TryGetValue(resolved, out var account)) return null;
        if (!account.Withdraw(amount)) return null;

        _outgoingTotals[resolved] += amount;
        _paymentCounter++;
        string paymentId = $"payment{_paymentCounter}";
        long cashbackTime = timestamp + MILLISECONDS_IN_A_DAY;
        int cashbackAmount = (amount * 2) / 100;

        _cashbackQueue.Enqueue((cashbackTime, resolved, cashbackAmount, paymentId), cashbackTime);
        _paymentStatus[resolved][paymentId] = "IN_PROGRESS";
        UpdateBalanceHistory(resolved, timestamp);
        return paymentId;
    }

    public string? GetPaymentStatus(long timestamp, string accountId, string paymentId)
    {
        ProcessCashbacks(timestamp);
        string resolved = ResolveAccount(accountId);
        return _paymentStatus.ContainsKey(resolved) && _paymentStatus[resolved].ContainsKey(paymentId)
            ? _paymentStatus[resolved][paymentId]
            : null;
    }

    public List<string> TopSpenders(long timestamp, int n)
    {
        ProcessCashbacks(timestamp);
        var mergedTotals = new Dictionary<string, int>();
        foreach (var kvp in _outgoingTotals)
        {
            string root = ResolveAccount(kvp.Key);
            if (!mergedTotals.ContainsKey(root)) mergedTotals[root] = 0;
            mergedTotals[root] += kvp.Value;
        }
        return mergedTotals
            .OrderByDescending(kv => kv.Value)
            .ThenBy(kv => kv.Key)
            .Take(n)
            .Select(kv => $"{kv.Key}({kv.Value})")
            .ToList();
    }

    public int? GetBalance(long timestamp, string accountId, long timeAt)
    {
        string resolved = ResolveAccount(accountId);
        if (!_accountHistory.ContainsKey(resolved)) return null;

        int? result = null;
        foreach (var (ts, bal) in _accountHistory[resolved])
        {
            if (ts <= timeAt) result = bal;
            else break;
        }
        return result;
    }

    public bool MergeAccounts(long timestamp, string acc1, string acc2)
    {
        string root1 = ResolveAccount(acc1);
        string root2 = ResolveAccount(acc2);
        if (root1 == root2 || !_accounts.ContainsKey(root1) || !_accounts.ContainsKey(root2)) return false;

        _accounts[root1].Deposit(_accounts[root2].GetBalance());
        _outgoingTotals[root1] += _outgoingTotals.GetValueOrDefault(root2);

        foreach (var kv in _paymentStatus[root2])
        {
            _paymentStatus[root1][kv.Key] = kv.Value;
        }
        foreach (var entry in _accountHistory[root2])
        {
            _accountHistory[root1].Add((timestamp, _accounts[root1].GetBalance()));
        }
        _mergeMap[root2] = root1;

        _accounts.Remove(root2);
        _outgoingTotals.Remove(root2);
        _paymentStatus.Remove(root2);
        _accountHistory.Remove(root2);
        return true;
    }

    private void ProcessCashbacks(long timestamp)
    {
        while (_cashbackQueue.Count > 0 && _cashbackQueue.Peek().timestamp <= timestamp)
        {
            var (cashbackTime, accountId, cashback, paymentId) = _cashbackQueue.Dequeue();
            string resolved = ResolveAccount(accountId);
            if (_accounts.ContainsKey(resolved))
            {
                _accounts[resolved].Deposit(cashback);
                _paymentStatus[resolved][paymentId] = "CASHBACK_RECEIVED";
                UpdateBalanceHistory(resolved, cashbackTime);
            }
        }
    }
}

// Account class remains unchanged and should support Withdraw, Deposit, and GetBalance methods as implemented earlier.

